SimExLite/SourceCalculators/GaussianSourceCalculator.py

```diff
@@ -0,0 +1,207 @@
+            "number_of_time_slices", comment="The number of time slices"
+        )
+        num_time_slices.add_interval(0, None, True)
+        num_time_slices.value = 12
+
+        param_z = parameters.new_parameter(
+            "z", comment="The position of the pulse in the beam direction", unit="m"
+        )
+        param_z.value = 100
+
+        self.parameters = parameters
+
+    def backengine(self):
+        from wpg.generators import build_gauss_wavefront
+        from wpg import Wavefront
+
+        # The rms of the amplitude distribution (Gaussian)
+        theta = self.parameters["divergence"].value_no_conversion.to("radian").magnitude
+        E_joule = (
+            self.parameters["photon_energy"].value_no_conversion.to("joule").magnitude
+        )
+        E_eV = self.parameters["photon_energy"].value_no_conversion.to("eV").magnitude
+        coherence_time = (
+            2.0
+            * np.pi
+            * hbar
+            / self.parameters["photon_energy_relative_bandwidth"].value
+            / E_joule
+        )
+        pulse_energy = self.parameters["pulse_energy"].value_no_conversion
+
+        beam_waist = 2.0 * hbar * c / theta / E_joule
+        wavelength = 1239.8e-9 / E_eV
+        rayleigh_length = np.pi * beam_waist**2 / wavelength
+
+        print("rayleigh_length =", rayleigh_length)
+
+        beam_diameter_fwhm = (
+            self.parameters["beam_diameter_fwhm"]
+            .value_no_conversion.to("meter")
+            .magnitude
+        )
+        beam_waist_radius = beam_diameter_fwhm / np.sqrt(2.0 * np.log(2.0))
+
+        # x-y range at beam waist.
+        range_xy = 30.0 * beam_waist_radius
+
+        # Set number of sampling points in x and y and number of temporal slices.
+        npoints = self.parameters["number_of_transverse_grid_points"].value
+        nslices = self.parameters["number_of_time_slices"].value
+
+        # Distance from source position.
+        z = self.parameters["z"].value_no_conversion.to("meter").magnitude
+
+        # Build wavefront
+        srwl_wf = build_gauss_wavefront(
+            npoints,
+            npoints,
+            nslices,
+            E_eV / 1.0e3,
+            -range_xy / 2,
+            range_xy / 2,
+            -range_xy / 2,
+            range_xy / 2,
+            coherence_time / np.sqrt(2),
+            beam_waist_radius / 2,
+            beam_waist_radius
+            / 2,  # Scaled such that fwhm comes out as demanded by parameters.
+            d2waist=z,
+            pulseEn=pulse_energy.to("joule").magnitude,
+            pulseRange=8.0,
+        )
+
+        # Correct radius of curvature.
+        Rx = Ry = z * np.sqrt(1.0 + (rayleigh_length / z) ** 2)
+
+        # Store on class.
+        srwl_wf.Rx = Rx
+        srwl_wf.Ry = Ry
+
+        key = self.output_keys[0]
+        filename = self.output_filenames[0]
+        output_data = self.output[key]
+
+        wavefront = Wavefront(srwl_wf)
+        wavefront.store_hdf5(filename)
```

Comment:
why is IO part of the backengine? i thought the new way of handling data is that backengine returns a Data object and IO is handled somewhere else or by the caller of the Calculator instance?

Reply:
Yes the backengine is returning the Data object, and here the Data object is a "file mapping" (line 195). The lines you are referring to is the file produced by the backengine. Maybe what you want is a "python object mapping", but it's not the case here. 
 
https://github.com/PaNOSC-ViNYL/SimEx-Lite/blob/40ec02a7f69ee1e5f545c05e844fc7bfcb1d6536/SimExLite/SourceCalculators/GaussianSourceCalculator.py#L192-L195